File: apps/cli/src/commands/tenants.ts
/**
 * `frick tenants list` — list rows in the `tenants` ledger.
 * `frick tenants create <id> [--display-name <name>]` — insert a tenant row.
 * `frick tenants set-push <id> --platform apns --p8 <file> --key-id ... --team-id ... --bundle-id ...` —
 *     wrap APNs credentials with `FRICK_PUSH_CRED_KEY` and store in `tenant_settings`.
 * `frick tenants set-push <id> --platform fcm --service-account <file>` —
 *     wrap FCM service-account JSON and store in `tenant_settings`.
 * `frick tenants set-push <id> --platform webpush --subject <mailto:|https:> --public-key <b64url> --private-key <pem-file>` —
 *     wrap Web Push VAPID credentials and store in `tenant_settings`.
 */
import { readFileSync } from "node:fs";
import {
  TenantAlreadyExistsError,
  saveApnsCredentials,
  saveFcmCredentials,
  saveWebPushCredentials,
} from "@fricken/server";
import type { ParsedArgs } from "../argv.js";
import { CliFailureError, CliUsageError } from "../errors.js";
import { contextFlagsFrom, loadConfig, openStore } from "../context.js";
import { emit, type OutputOptions } from "../output.js";
import { requireString } from "../argv.js";

export async function tenantsCommand(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const sub = parsed.positionals[0];
  if (sub === "list") return tenantsList(parsed, out);
  if (sub === "create") return tenantsCreate(parsed, out);
  if (sub === "set-push") return tenantsSetPush(parsed, out);
  throw new CliUsageError(`Unknown tenants subcommand: ${sub ?? "<missing>"}`, {
    expected: ["list", "create", "set-push"],
  });
}

async function tenantsList(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const config = loadConfig(contextFlagsFrom(parsed.flags));
  const store = openStore(config);
  try {
    const includeArchived = parsed.flags["include-archived"] === true;
    const rows = await store.tenants.list(includeArchived);
    emit({ tenants: rows }, out);
    return 0;
  } finally {
    store.close();
  }
}

function requireFlag(flags: Record<string, string | boolean>, key: string): string {
  const value = requireString(flags, key);
  if (!value) {
    throw new CliUsageError(`Missing required --${key} flag`);
  }
  return value;
}

async function tenantsSetPush(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const tenantId = parsed.positionals[1];
  if (!tenantId) {
    throw new CliUsageError("frick tenants set-push requires a tenant id positional argument");
  }
  const platform = requireFlag(parsed.flags, "platform");
  const config = loadConfig(contextFlagsFrom(parsed.flags));
  const store = openStore(config);
  try {
    if (platform === "apns") {
      const p8Path = requireFlag(parsed.flags, "p8");
      const keyId = requireFlag(parsed.flags, "key-id");
      const teamId = requireFlag(parsed.flags, "team-id");
      const bundleId = requireFlag(parsed.flags, "bundle-id");
      const useSandbox = parsed.flags.sandbox === true;
      const privateKeyPem = readFileSync(p8Path, "utf8");
      const result = await saveApnsCredentials(store.tenantSettings, tenantId, {
        keyId,
        teamId,
        bundleId,
        privateKeyPem,
        useSandbox,
      });
      if (!result.ok) {
        throw new CliFailureError(result.error.code, result.error.message, { tenantId });
      }
      emit({ ok: true, tenantId, platform: "apns" }, out);
      return 0;
    }
    if (platform === "fcm") {
      const path = requireFlag(parsed.flags, "service-account");
      const raw = readFileSync(path, "utf8");
      let parsedJson: { project_id?: string; client_email?: string; private_key?: string; token_uri?: string };
      try {
        parsedJson = JSON.parse(raw) as typeof parsedJson;
      } catch {
        throw new CliFailureError(
          "tenants.setPush.invalidServiceAccount",
          "Service-account file is not valid JSON",
          { path },
        );
      }
      if (!parsedJson.project_id || !parsedJson.client_email || !parsedJson.private_key) {
        throw new CliFailureError(
          "tenants.setPush.invalidServiceAccount",
          "Service-account file missing project_id, client_email, or private_key",
          { path },
        );
      }
      const result = await saveFcmCredentials(store.tenantSettings, tenantId, {
        projectId: parsedJson.project_id,
        clientEmail: parsedJson.client_email,
        privateKey: parsedJson.private_key,
        ...(parsedJson.token_uri ? { tokenUri: parsedJson.token_uri } : {}),
      });
      if (!result.ok) {
        throw new CliFailureError(result.error.code, result.error.message, { tenantId });
      }
      emit({ ok: true, tenantId, platform: "fcm" }, out);
      return 0;
    }
    if (platform === "webpush") {
      const subject = requireFlag(parsed.flags, "subject");
      if (!subject.startsWith("mailto:") && !subject.startsWith("https://")) {
        throw new CliFailureError(
          "tenants.setPush.invalidVapidSubject",
          "Web Push --subject must be a mailto: or https:// URI",
          { subject },
        );
      }
      const publicKey = requireFlag(parsed.flags, "public-key");
      const privateKeyPath = requireFlag(parsed.flags, "private-key");
      const privateKey = readFileSync(privateKeyPath, "utf8");
      const result = await saveWebPushCredentials(store.tenantSettings, tenantId, {
        subject,
        publicKey,
        privateKey,
      });
      if (!result.ok) {
        throw new CliFailureError(result.error.code, result.error.message, { tenantId });
      }
      emit({ ok: true, tenantId, platform: "webpush" }, out);
      return 0;
    }
    throw new CliUsageError(`Unsupported --platform: ${platform}`, {
      expected: ["apns", "fcm", "webpush"],
    });
  } finally {
    store.close();
  }
}

async function tenantsCreate(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const tenantId = parsed.positionals[1];
  if (!tenantId) {
    throw new CliUsageError("frick tenants create requires a tenant id positional argument");
  }
  const displayName = requireString(parsed.flags, "display-name");
  const config = loadConfig(contextFlagsFrom(parsed.flags));
  const store = openStore(config);
  try {
    const row = await store.tenants.create(tenantId, displayName);
    emit({ ok: true, tenant: row }, out);
    return 0;
  } catch (error) {
    if (error instanceof TenantAlreadyExistsError) {
      throw new CliFailureError("tenants.exists", error.message, { tenantId });
    }
    throw error;
  } finally {
    store.close();
  }
}

File: crates/frick-cli/src/commands/tenants.rs
//! `frick tenants list|create|set-push`
//! (ported from `apps/cli/src/commands/tenants.ts`).
//!
//! `list` / `create` go through `store.tenants()`. `set-push` wraps APNs / FCM /
//! Web Push credentials with `FRICK_PUSH_CRED_KEY` and stores them in
//! `tenant_settings`, using the `frick_server::push::credentials` helpers — the
//! same sealing the server's credential loader can decrypt.

use std::fs;
use std::time::{SystemTime, UNIX_EPOCH};

use frick_server::push::credentials::{
    ApnsCredentials, FcmCredentials, ProcessCredentialEnv, WebPushCredentials,
    save_apns_credentials, save_fcm_credentials, save_web_push_credentials,
};
use serde_json::json;

use crate::argv::ParsedArgs;
use crate::context::{context_flags_from, load_config, open_store};
use crate::errors::{CliError, EXIT_OK};
use crate::output::Output;

fn now_ms() -> i64 {
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |d| d.as_millis());
    i64::try_from(millis).unwrap_or(i64::MAX)
}

/// `tenantsCommand`.
pub async fn tenants_command(parsed: &ParsedArgs, out: &mut Output<'_>) -> Result<i32, CliError> {
    match parsed.positional(0) {
        Some("list") => tenants_list(parsed, out).await,
        Some("create") => tenants_create(parsed, out).await,
        Some("set-push") => tenants_set_push(parsed, out).await,
        other => Err(CliError::usage_with(
            format!(
                "Unknown tenants subcommand: {}",
                other.unwrap_or("<missing>")
            ),
            json!({ "expected": ["list", "create", "set-push"] }),
        )),
    }
}

async fn tenants_list(parsed: &ParsedArgs, out: &mut Output<'_>) -> Result<i32, CliError> {
    let config = load_config(&context_flags_from(parsed))?;
    let store = open_store(&config).await?;
    let include_archived = parsed.flag_bool_present("include-archived");
    let rows = store
        .tenants()
        .list(include_archived)
        .await
        .map_err(|err| CliError::failure("cli.store", err.to_string()))?;
    let rows: Vec<_> = rows
        .iter()
        .map(|row| {
            json!({
                "tenantId": row.tenant_id,
                "displayName": row.display_name,
                "createdAt": row.created_at,
                "archivedAt": row.archived_at,
            })
        })
        .collect();
    out.emit(&json!({ "tenants": rows }));
    Ok(EXIT_OK)
}

async fn tenants_create(parsed: &ParsedArgs, out: &mut Output<'_>) -> Result<i32, CliError> {
    let Some(tenant_id) = parsed.positional(1) else {
        return Err(CliError::usage(
            "frick tenants create requires a tenant id positional argument",
        ));
    };
    let display_name = parsed.flag_str("display-name");
    let config = load_config(&context_flags_from(parsed))?;
    let store = open_store(&config).await?;
    match store
        .tenants()
        .create(tenant_id, display_name, now_ms())
        .await
    {
        Ok(row) => {
            out.emit(&json!({
                "ok": true,
                "tenant": {
                    "tenantId": row.tenant_id,
                    "displayName": row.display_name,
                    "createdAt": row.created_at,
                    "archivedAt": row.archived_at,
                },
            }));
            Ok(EXIT_OK)
        }
        Err(err) => {
            let message = err.to_string();
            if message == frick_store::stores::tenant::tenant_already_exists_message(tenant_id) {
                Err(CliError::failure_with(
                    "tenants.exists",
                    message,
                    json!({ "tenantId": tenant_id }),
                ))
            } else {
                Err(CliError::failure("cli.store", message))
            }
        }
    }
}

fn require_flag<'a>(parsed: &'a ParsedArgs, key: &str) -> Result<&'a str, CliError> {
    match parsed.flag_str(key) {
        Some(value) if !value.is_empty() => Ok(value),
        _ => Err(CliError::usage(format!("Missing required --{key} flag"))),
    }
}

async fn tenants_set_push(parsed: &ParsedArgs, out: &mut Output<'_>) -> Result<i32, CliError> {
    let Some(tenant_id) = parsed.positional(1).map(ToString::to_string) else {
        return Err(CliError::usage(
            "frick tenants set-push requires a tenant id positional argument",
        ));
    };
    let platform = require_flag(parsed, "platform")?.to_string();
    let config = load_config(&context_flags_from(parsed))?;
    let store = open_store(&config).await?;
    let env = ProcessCredentialEnv;

    match platform.as_str() {
        "apns" => set_push_apns(parsed, &store, &tenant_id, &env, out).await,
        "fcm" => set_push_fcm(parsed, &store, &tenant_id, &env, out).await,
        "webpush" => set_push_webpush(parsed, &store, &tenant_id, &env, out).await,
        other => Err(CliError::usage_with(
            format!("Unsupported --platform: {other}"),
            json!({ "expected": ["apns", "fcm", "webpush"] }),
        )),
    }
}

fn cred_error(
    tenant_id: &str,
) -> impl Fn(frick_server::push::credentials::PushCredentialError) -> CliError + '_ {
    move |err| {
        CliError::failure_with(
            err.code.as_str(),
            err.message,
            json!({ "tenantId": tenant_id }),
        )
    }
}

async fn set_push_apns(
    parsed: &ParsedArgs,
    store: &frick_store::FrickStore,
    tenant_id: &str,
    env: &ProcessCredentialEnv,
    out: &mut Output<'_>,
) -> Result<i32, CliError> {
    let p8_path = require_flag(parsed, "p8")?;
    let creds = ApnsCredentials {
        key_id: require_flag(parsed, "key-id")?.to_string(),
        team_id: require_flag(parsed, "team-id")?.to_string(),
        bundle_id: require_flag(parsed, "bundle-id")?.to_string(),
        private_key_pem: read_file(p8_path)?,
        use_sandbox: parsed.flag_bool_present("sandbox"),
    };
    save_apns_credentials(store.tenant_settings(), tenant_id, &creds, env, now_ms())
        .await
        .map_err(cred_error(tenant_id))?;
    out.emit(&json!({ "ok": true, "tenantId": tenant_id, "platform": "apns" }));
    Ok(EXIT_OK)
}

async fn set_push_fcm(
    parsed: &ParsedArgs,
    store: &frick_store::FrickStore,
    tenant_id: &str,
    env: &ProcessCredentialEnv,
    out: &mut Output<'_>,
) -> Result<i32, CliError> {
    let path = require_flag(parsed, "service-account")?;
    let raw = read_file(path)?;
    let parsed_json: serde_json::Value = serde_json::from_str(&raw).map_err(|_| {
        CliError::failure_with(
            "tenants.setPush.invalidServiceAccount",
            "Service-account file is not valid JSON",
            json!({ "path": path }),
        )
    })?;
    let project_id = parsed_json.get("project_id").and_then(|v| v.as_str());
    let client_email = parsed_json.get("client_email").and_then(|v| v.as_str());
    let private_key = parsed_json.get("private_key").and_then(|v| v.as_str());
    let (Some(project_id), Some(client_email), Some(private_key)) =
        (project_id, client_email, private_key)
    else {
        return Err(CliError::failure_with(
            "tenants.setPush.invalidServiceAccount",
            "Service-account file missing project_id, client_email, or private_key",
            json!({ "path": path }),
        ));
    };
    let creds = FcmCredentials {
        project_id: project_id.to_string(),
        client_email: client_email.to_string(),
        private_key: private_key.to_string(),
        token_uri: parsed_json
            .get("token_uri")
            .and_then(|v| v.as_str())
            .map(ToString::to_string),
    };
    save_fcm_credentials(store.tenant_settings(), tenant_id, &creds, env, now_ms())
        .await
        .map_err(cred_error(tenant_id))?;
    out.emit(&json!({ "ok": true, "tenantId": tenant_id, "platform": "fcm" }));
    Ok(EXIT_OK)
}

async fn set_push_webpush(
    parsed: &ParsedArgs,
    store: &frick_store::FrickStore,
    tenant_id: &str,
    env: &ProcessCredentialEnv,
    out: &mut Output<'_>,
) -> Result<i32, CliError> {
    let subject = require_flag(parsed, "subject")?.to_string();
    if !subject.starts_with("mailto:") && !subject.starts_with("https://") {
        return Err(CliError::failure_with(
            "tenants.setPush.invalidVapidSubject",
            "Web Push --subject must be a mailto: or https:// URI",
            json!({ "subject": subject }),
        ));
    }
    let public_key = require_flag(parsed, "public-key")?.to_string();
    let private_key_path = require_flag(parsed, "private-key")?;
    let creds = WebPushCredentials {
        subject,
        public_key,
        private_key: read_file(private_key_path)?,
    };
    save_web_push_credentials(store.tenant_settings(), tenant_id, &creds, env, now_ms())
        .await
        .map_err(cred_error(tenant_id))?;
    out.emit(&json!({ "ok": true, "tenantId": tenant_id, "platform": "webpush" }));
    Ok(EXIT_OK)
}

fn read_file(path: &str) -> Result<String, CliError> {
    fs::read_to_string(path)
        .map_err(|err| CliError::failure("cli.io", format!("Could not read {path}: {err}")))
}
